packages/contracts/src/contracts/current/forwarder/MixinForwarderExpectedResults.sol

```diff
@@ -0,0 +1,130 @@
+pragma solidity ^0.4.24;
+pragma experimental ABIEncoderV2;
+
+import "./MixinForwarderCore.sol";
+
+contract MixinForwarderExpectedResults is MixinForwarderCore {
+
+    /// @dev Simulates the 0x Exchange fillOrder validation and calculations, without performing any state changes.
+    /// @param order An Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function calculateFillOrderFillResults(
+        Order memory order,
+        uint256 takerAssetFillAmount
+    )
+        public
+        view
+        returns (Exchange.FillResults memory fillResults)
+    {
+        // Exchange.OrderInfo memory orderInfo = EXCHANGE.getOrderInfo(order);
+        fillResults = EXCHANGE.calculateFillResults(order, takerAssetFillAmount);
+        return fillResults;
+    }
+
+    /// @dev Calculates a FillResults total for selling takerAssetFillAmount over all orders. 
+    ///      Including the fees required to be paid. 
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param takerAssetFillAmount A number representing the amount of this order to fill.
+    /// @return totalFillResults Amounts filled and fees paid by maker and taker.
+    function calculateMarketSellFillResults(
+        Order[] memory orders,
+        uint256 takerAssetFillAmount
+    )
+        internal
+        view
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        for (uint256 i = 0; i < orders.length; i++) {
+            require(
```

Comment:
Do we still need this check after optimizing the calldata?

Reply:
Good catch I think we can completely drop this